### PR TITLE
Add logging to common covenant events

### PR DESCRIPTION
**Pull request by checkymander**

### CODE REVIEW

**Comment by checkymander** [comment]
merged

**Review by cobbr — CHANGES_REQUESTED** [review]

**Comment by checkymander** [comment]
Added a new request restructuring the loggingservice to be more in line with the way you envisioned it. I've also added a command line option to allow users to also log command output, but I still need to test it some more. 

Here's an example of the log output
```
2020-11-07 19:28:28.6904|TRACE|Covenant.Core.Common|Listener|Update|2|38097fdead|df56d480a1|0.0.0.0|80|192.168.86.22|80|1|HTTP|Active
2020-11-07 19:28:44.0552|TRACE|Covenant.Core.Common|Listener|Delete|2|38097fdead|df56d480a1|0.0.0.0|80|192.168.86.22|80|1|HTTP|Stopped
2020-11-07 19:28:49.5949|TRACE|Covenant.Core.Common|CovenantUser|Create|eb82bb0b-f25a-4494-9955-d8a98bc36e9c|D62D7AB5-08EB-DF74-A23F-AE44377D169B
2020-11-07 19:28:49.8285|TRACE|Covenant.Core.Common|Listener|Create|3|968f9a0af4|3e080e3493|0.0.0.0|80|192.168.86.22|80|1|HTTP|Uninitialized
2020-11-07 19:28:50.4956|TRACE|Covenant.Core.Common|Listener|Update|3|968f9a0af4|3e080e3493|0.0.0.0|80|192.168.86.22|80|1|HTTP|Active
2020-11-07 19:28:50.5825|TRACE|Covenant.Core.Common|Listener|Update|3|968f9a0af4|3e080e3493|0.0.0.0|80|192.168.86.22|80|1|HTTP|Active
2020-11-07 19:33:57.1464|TRACE|Covenant.Core.Common|Listener|Update|3|968f9a0af4|3e080e3493|0.0.0.0|80|192.168.86.22|80|1|HTTP|Active
2020-11-07 19:35:02.8773|TRACE|Covenant.Core.Common|Grunt|Create|9|defd6a3dae|DESKTOP-CQRINT5|Medium|192.168.86.22|DESKTOP-CQRINT5
2020-11-07 19:35:35.4023|TRACE|Covenant.Core.Common|GruntCommand|Create|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|3|ChangeDirectory C:\|
2020-11-07 19:35:39.0044|TRACE|Covenant.Core.Common|GruntCommand|Update|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|3|ChangeDirectory C:\|
2020-11-07 19:35:39.0230|TRACE|Covenant.Core.Common|GruntTasking|Create|3|01451fa1b9|9|17|Uninitialized
2020-11-07 19:35:39.0553|TRACE|Covenant.Core.Common|GruntCommand|Update|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|3|ChangeDirectory C:\|
2020-11-07 19:35:41.2752|TRACE|Covenant.Core.Common|GruntTasking|Update|3|01451fa1b9|9|17|Tasked
2020-11-07 19:35:46.7399|TRACE|Covenant.Core.Common|GruntTasking|Update|3|01451fa1b9|9|17|Completed
2020-11-07 19:35:55.1106|TRACE|Covenant.Core.Common|GruntCommand|Create|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|4|ls|
2020-11-07 19:35:55.1546|TRACE|Covenant.Core.Common|GruntCommand|Update|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|4|ls|
2020-11-07 19:35:58.3080|TRACE|Covenant.Core.Common|GruntCommand|Create|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|5|dir|
2020-11-07 19:35:58.3505|TRACE|Covenant.Core.Common|GruntCommand|Update|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|5|dir|
2020-11-07 19:36:51.2798|TRACE|Covenant.Core.Common|GruntCommand|Create|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|6|SharpShell /code:"dir"|
2020-11-07 19:36:51.6815|TRACE|Covenant.Core.Common|GruntCommand|Update|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|6|SharpShell /code:"dir"|
2020-11-07 19:36:51.6897|TRACE|Covenant.Core.Common|GruntTasking|Create|4|bde27956a7|9|23|Aborted
2020-11-07 19:37:20.0445|TRACE|Covenant.Core.Common|GruntCommand|Create|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|7|GetCurrentDirectory|
2020-11-07 19:37:20.2726|TRACE|Covenant.Core.Common|GruntCommand|Update|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|7|GetCurrentDirectory|
2020-11-07 19:37:20.2811|TRACE|Covenant.Core.Common|GruntTasking|Create|5|1fe3cbd4cb|9|16|Uninitialized
2020-11-07 19:37:20.3327|TRACE|Covenant.Core.Common|GruntCommand|Update|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|7|GetCurrentDirectory|
2020-11-07 19:37:24.0579|TRACE|Covenant.Core.Common|GruntTasking|Update|5|1fe3cbd4cb|9|16|Tasked
2020-11-07 19:37:29.3613|TRACE|Covenant.Core.Common|GruntTasking|Update|5|1fe3cbd4cb|9|16|Completed
2020-11-07 19:39:28.4313|TRACE|Covenant.Core.Common|GruntCommand|Create|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|8|ReadTextFile /path:"C:\Users\scott\Desktop\test.html"|
2020-11-07 19:39:28.6645|TRACE|Covenant.Core.Common|GruntCommand|Update|c|b8249947-efd6-4502-9192-c3ebb690cd3e|9|8|ReadTextFile /path:"C:\Users\scott\Desktop\test.html"|
2020-11-07 19:39:28.6734|TRACE|Covenant.Core.Common|GruntTasking|Create|6|75ecca7147|9|18|Uninitialized
2020-11-07 19:39:32.2344|TRACE|Covenant.Core.Common|GruntTasking|Update|6|75ecca7147|9|18|Tasked
2020-11-07 19:39:37.5639|TRACE|Covenant.Core.Common|GruntTasking|Update|6|75ecca7147|9|18|Completed
```

There's some bugs I can definitely see for the ELK guys when parsing logs which I need to think about if you have any thoughts (e.g. if a command line or output has a | in it, it would probably break their log parser)

Let me know what you think and if there are any changes that I can make!

File: Covenant/Core/LoggingService.cs [hunk 1]
```diff
@@ -0,0 +1,431 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (checkymander) [reply Covenant/Core/LoggingService.cs:1]: Finished.

File: Covenant/Core/LoggingService.cs [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+            {
+                _logger.Info($"[OnCreateCovenantUser]|{covenantUser.Id}|{covenantUser.NormalizedUserName}");
+            };
+            //this.OnEditCovenantUser += (sender, covenantUser) =>
+            //{
+            //    _logger.Info($"[OnEditCovenantUser] {covenantUser.Id} - {covenantUser.NormalizedUserName}");
+            //};
+            this.OnDeleteCovenantUser += (sender, covenantUser) =>
+            {
+                _logger.Info($"[OnDeleteCovenantUser]|UserID:{covenantUser.Id}|UserName:{covenantUser.NormalizedUserName}");
```
[the remark this answers is not in the document]
Comment (checkymander) [reply Covenant/Core/LoggingService.cs]: done.

File: Covenant/Core/LoggingService.cs [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+using Covenant.Models.Launchers;
+using Covenant.Models.Grunts;
+using Covenant.Models.Indicators;
+using NLog;
+
+namespace Covenant.Core
+{
+    public interface ICovenantUserLoggingService
+    {
+        event EventHandler<CovenantUser> OnCreateCovenantUser;
```
[the remark this answers is not in the document]
Comment (checkymander) [reply Covenant/Core/LoggingService.cs]: done.

File: Covenant/Core/LoggingService.cs [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+            //};
+            this.OnDeleteCovenantUser += (sender, covenantUser) =>
+            {
+                _logger.Info($"[OnDeleteCovenantUser]|UserID:{covenantUser.Id}|UserName:{covenantUser.NormalizedUserName}");
+            };
+
+
+            this.OnCreateIdentityRole += (sender, role) =>
+            {
+                _logger.Info($"[OnCreateIdentityRole]|RoleID:{role.Id}|RoleName:{role.Name} ");
```
[the remark this answers is not in the document]
Comment (checkymander) [reply Covenant/Core/LoggingService.cs]: done.

**Pull request merged**